src/plugins/workspace/server/workspace_client.ts

```diff
@@ -39,6 +39,43 @@ import {
   WORKSPACE_UPDATE_APP_ID,
 } from '../common/constants';
 
+const validatePermissionModesCombinations = [
+  [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.Read], // Read
+  [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Read], // Write
+  [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Write], // Admin
+];
+
+const isValidatePermissionModesCombination = (permissionModes: string[]) =>
+  validatePermissionModesCombinations.some(
+    (combination) =>
+      combination.length === permissionModes.length &&
+      combination.every((mode) => permissionModes.includes(mode))
+  );
+const isValidatePermissions = (permissions: Permissions) => {
```

Comment:
If we move `convertToACL` to `workspace_client`, and passing the raw permissions(type is `WorkspaceRoutePermissionItem[]`) to create/update, perhaps we don't need to convert the data here? It would be much easier to validate a flatten array of `WorkspaceRoutePermissionItem[]`.

Not only make it easier to validate the permission mode combinations, but also make it easier to validate the user/group duplications here as well.

Reply:
That's a good idea. The `WorkspaceRoutePermissionItem[]` in route level was only used in the front end page. I kept the `ACL` parameter in case of passing `ACL` to workspace client directly. From current implementation, the `create` or `update` method only be used in workspace routes. I think we can change to parameter type to `WorkspaceRoutePermissionItem[]`.

Reply:
we also can rename `WorkspaceRoutePermissionItem` to not include word `Route` so that it's more generic,